File: src/vehicle/rental/system/Invoice/InvoicePrinter.java
package vehicle.rental.system.Invoice;

import java.io.PrintStream;
import java.math.BigDecimal;
import java.math.RoundingMode;

public class InvoicePrinter {
    private final PrintStream printer;

    private InvoicePrinter(PrintStream printer) {
        this.printer = printer;
    }

    public static InvoicePrinter getInstance() {return new InvoicePrinter(System.out);}

    public void printInvoice(final Invoice invoice) {

        printer.println("XXXXXXXXXX");
        printer.println("Date: " + invoice.date());
        printer.println("Customer Name: " + invoice.customerName());
        printer.println("Rented vehicle: " + invoice.vehicleBrand() + " " + invoice.vehicleModel() + "\n");
        printer.println("Reservation start date: " + invoice.startDate());
        printer.println("Reservation end date: " + invoice.endDate());
        printer.println("Reserved rental days: " + invoice.reservedDays() + " days\n");
        printer.println("Actual Return date: " + invoice.actualReturnDate());
        printer.println("Actual rental days: " + invoice.actualDays() + " days\n");
        printer.println("Rental cost per day: $" + format(invoice.dailyRentalCost()));
        printer.println("Insurance per day: $" + format(invoice.dailyInsuranceCost()));
        printer.println("Total rent: $" + format(invoice.totalRentalCost()));
        printer.println("Total insurance: $" + format(invoice.totalInsuranceCost()));
        printer.println("Total: $" + format(invoice.totalInsuranceCost().add(invoice.totalRentalCost())));
        printer.println("XXXXXXXXXX");
    }
    private static String format(final BigDecimal value) {
        return value.setScale(2, RoundingMode.DOWN).toPlainString();
    }

}

File: src/Invoice/InvoicePrinter.cs
namespace VehicleRentalSystem.Invoice;

public class InvoicePrinter
{
    private readonly TextWriter _writer;

    private InvoicePrinter(TextWriter writer)
    {
        _writer = writer;
    }

    public static InvoicePrinter GetInstance() => new(Console.Out);

    public void PrintInvoice(Invoice invoice)
    {
        _writer.WriteLine("XXXXXXXXXX");
        _writer.WriteLine($"Date: {invoice.Date}");
        _writer.WriteLine($"Customer Name: {invoice.CustomerName}");
        _writer.WriteLine($"Rented vehicle: {invoice.VehicleBrand} {invoice.VehicleModel}\n");
        _writer.WriteLine($"Reservation start date: {invoice.StartDate}");
        _writer.WriteLine($"Reservation end date: {invoice.EndDate}");
        _writer.WriteLine($"Reserved rental days: {invoice.ReservedDays} days\n");
        _writer.WriteLine($"Actual Return date: {invoice.ActualReturnDate}");
        _writer.WriteLine($"Actual rental days: {invoice.ActualDays} days\n");
        _writer.WriteLine($"Rental cost per day: ${Format(invoice.DailyRentalCost)}");
        _writer.WriteLine($"Insurance per day: ${Format(invoice.DailyInsuranceCost)}");
        _writer.WriteLine($"Total rent: ${Format(invoice.TotalRentalCost)}");
        _writer.WriteLine($"Total insurance: ${Format(invoice.TotalInsuranceCost)}");
        _writer.WriteLine($"Total: ${Format(invoice.TotalInsuranceCost + invoice.TotalRentalCost)}");
        _writer.WriteLine("XXXXXXXXXX");
    }

    private static string Format(decimal value) => Math.Truncate(value * 100) / 100 is var v ? v.ToString("F2") : "0.00";
}
